src/coreclr/jit/optcse.cpp

```diff
@@ -2780,18 +2794,71 @@ void CSE_HeuristicRL::GetFeatures(CSEdsc* cse, double* features)
     features[13] = booleanScale * (isMinCost & isLiveAcrossCall);
 
     // Is any CSE tree for this candidate marked GTF_MAKE_CSE (hoisting)
+    // Also gather data for "distance" metric.
     //
-    bool isMakeCse = false;
+    const unsigned numBBs            = m_pCompiler->fgBBcount;
+    bool           isMakeCse         = false;
+    unsigned       minPostorderNum   = numBBs;
+    unsigned       maxPostorderNum   = 0;
+    BasicBlock*    minPostorderBlock = nullptr;
+    BasicBlock*    maxPostorderBlock = nullptr;
     for (treeStmtLst* treeList = cse->csdTreeList; treeList != nullptr && !isMakeCse; treeList = treeList->tslNext)
     {
-        isMakeCse = ((treeList->tslTree->gtFlags & GTF_MAKE_CSE) != 0);
+        BasicBlock* const treeBlock    = treeList->tslBlock;
+        unsigned          postorderNum = treeBlock->bbPostorderNum;
+        if (postorderNum < minPostorderNum)
+        {
+            minPostorderNum   = postorderNum;
+            minPostorderBlock = treeBlock;
+        }
+
+        if (postorderNum > maxPostorderNum)
+        {
+            maxPostorderNum   = postorderNum;
+            maxPostorderBlock = treeBlock;
+        }
+
+        isMakeCse |= ((treeList->tslTree->gtFlags & GTF_MAKE_CSE) != 0);
     }
+    const unsigned blockSpread = maxPostorderNum - minPostorderNum;
+
     features[14] = booleanScale * isMakeCse;
 
     // Locals data
     //
     features[15] = cse->numDistinctLocals;
     features[16] = cse->numLocalOccurrences;
+
+    // More
+    //
+    features[17] = booleanScale * ((cse->csdTree->gtFlags & GTF_CALL) != 0);
+    features[18] = deMinimusAdj + log(max(deMinimis, cse->csdUseCount * cse->csdUseWtCnt));
+    features[19] = deMinimusAdj + log(max(deMinimis, cse->numLocalOccurrences * cse->csdUseWtCnt));
+    features[20] = booleanScale * ((double)(blockSpread) / numBBs);
+
+    const bool isContainable = cse->csdTree->OperIs(GT_ADD, GT_NOT, GT_MUL, GT_LSH);
```

Comment:
shouldn't such nodes be marked with `GTF_ADDRMODE_NO_CSE` by this point? Assuming you mean they could be contained as part of addressing modes

Reply:
This stage of CSE will never see anything marked `GTF_DONT_CSE`.

This is a more general guess as to whether the operation can be subsumed by its parent, in particular I saw an AND(NOT ...) turn into `andn` so CSE of the NOT is not always an improvement.